### DISCUSSION TITLE
how to get the usb serial port?

### DISCUSSION THREAD

**Discussion by DylanJMas** [xblax/flashforge_ad5m_klipper_mod#199] [Q&A]
Hello, 

I'm been trying to get my palette 2 working and I need the usb serial port to put it in the config. Its different then the raspberry pi usually to get the serial port you type in ls (/dev/serial/by-id/*) but this is set up differently.

So does anyone know or tell me how to get it.

 Thank you :')

**Comment by ballaswag** [comment]
Maybe this will work for you https://github.com/ballaswag/creality_k1_klipper_mod/blob/master/build_overlay/common/usr/libexec/symlink-serial-dev.sh

I needed the same thing from this system for the K1 and created the above script for mdev.

So you would place that script in /usr/libexec, then edit you /etc/mdev.conf to call it https://github.com/ballaswag/creality_k1_klipper_mod/blob/master/build_overlay/common/etc/mdev.conf#L24

My devices are added a ttyACM[0-9], your USB serial devices might be different.

**Reply by consp** [reply]
A [m/u]dev rule is a good point, since there was talk of installing cutters/multi-filament devices this is definitely a good thing to have.

As long as it's consistent it should not be too much of an issue and by the looks of it, it's a lot better than tty[dev][num] naming. I know of some FTDI serial chips which only work as tty once configured and they do not show up as tty's unless properly initialized after initial insertion so that won't work for those. Might not be an issue anyway since they are unlikely to be in consumer devices anyway due to cost. Most will use one of the cheaper CH3xx chips or Microchip USB to UART controllers which should work out of the box.

If you have multiple serial devices it might be good to check the load order since it might also screw up the klipper communication to the two MCU's controlling the printer if the order is changed (also happens with additional usb camera's with buttons with respect to the libev event devices).

**Reply by DylanJMas** [reply]
> A [m/u]dev rule is a good point, since there was talk of installing cutters/multi-filament devices this is definitely a good thing to have.
> 
> As long as it's consistent it should not be too much of an issue and by the looks of it, it's a lot better than tty[dev][num] naming. I know of some FTDI serial chips which only work as tty once configured and they do not show up as tty's unless properly initialized after initial insertion so that won't work for those. Might not be an issue anyway since they are unlikely to be in consumer devices anyway due to cost. Most will use one of the cheaper CH3xx chips or Microchip USB to UART controllers which should work out of the box.
> 
> If you have multiple serial devices it might be good to check the load order since it might also screw up the klipper communication to the two MCU's controlling the printer if the order is changed (also happens with additional usb camera's with buttons with respect to the libev event devices).




Is it possible to connect  a USB device other then a camera to the current firmware? RN I'm just copying and pasting random dir like dev/tty10 then typing in PALETTE_CONNECT, Im not getting any luck RN :'(

**Reply by consp** [reply]
@xblax might know the answer to this, I'm not that familiar with the current mdev configuration for uarts/tty's.

**Comment by xblax** [comment]
I gave this a try on my 5M. Due to the lack of USB Serial devices at hand I just connected an Arduino Nano via USB, which should show up as USB Serial Device with CH3xx.

On my desktop, this is registered as /dev/ttyUSB0 but nothing happens on the 5M. Turns out out, the Kernel does not have CONFIG_USB_SERIAL - but we might be able to compile that as a module.

The Kernel has CONFIG_USB_ACM=y but only some USB Serial devices use the CDC ACM interface to provide a serial port (actually meant for modems).

**Reply by xblax** [reply]
Yep, so conceptually this is working. However, we don't want to load all the possible USB Serial driver modules as that would waste memory.

We need to ship the included kernel modules as proper module tree so that modprobe can load them and configure mdev to handle the hotplug events and trigger modprobe to load necessary driver modules on demand (it's not too difficult).

**Reply by cfelicio** [reply]
Same situation for me, I'd love to be able to use the Palette 2 with this printer, but atm the module is not present on the kernel. I tried compiling usbserial.ko and ftdi_sio.ko myself, but not successful. Let me know if there is any more info on this so I can give it another try. Thanks! :-)
